src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala

```diff
@@ -3,114 +3,81 @@ package viper.silver.plugin.standard.inline
+    * @return The expression with expanded predicates and the expandable precondition and postcondition predicates
     */
-  private[this] def expandPredicates(expr: Exp, method: Method, program: Program, preds: Set[String]): Exp = {
+  private[this] def expandPredicates(expr: Exp, method: Method, program: Program, cond: String => Boolean): Exp = {
+    val expandablePredicates = mutable.Set[String]()
     ViperStrategy.Context[Set[String]]({
       case exp@(PredicateAccessPredicate(pred, perm), ctxt) =>
-        val isInUnfolding = ctxt.parentOption.exists({_.isInstanceOf[Unfolding]})
-        if (preds(pred.predicateName) && !isInUnfolding) {
-          val maybePredBody = pred.predicateBody(program, ctxt.c)
-          (propagatePermission(maybePredBody, perm).get, ctxt)
+        val isInUnfolding = ctxt.parentOption.exists(_.isInstanceOf[Unfolding])
+        if (cond(pred.predicateName) && !isInUnfolding) {
+          expandablePredicates += pred.predicateName
+          val optPredBody = pred.predicateBody(program, ctxt.c)
+          (propagatePermission(optPredBody, perm).get, ctxt)
         } else exp
-      case (quant: QuantifiedExp, ctxt) =>
-        (quant, ctxt.updateContext(ctxt.c ++ quant.scopedDecls.map { _.name }.toSet))
-    }, method.scopedDecls.map { _.name }.toSet, Traverse.TopDown)
+      case (scope: Scope, ctxt) =>
+        (scope, ctxt.updateContext(ctxt.c ++ scope.scopedDecls.map(_.name).toSet))
+    }, method.scopedDecls.map(_.name).toSet, Traverse.TopDown)
       .execute[Exp](expr)
   }
 
   /**
     * Replaces unfolding expressions by their bodies if the unfolded predicate had been expanded
     *
     * @param expr The expression whose unfoldings may be substituted
-    * @param preds A set of the string names of the predicates that have been expanded
+    * @param cond The condition a predicate must satisfy to be un-unfolding-ed
     * @return The expression with unfoldings possibly substituted
     */
-  private[this] def removeUnfoldings(expr: Exp, preds: Set[String]): Exp = {
+  private[this] def removeUnfoldings(expr: Exp, cond: String => Boolean): Exp = {
     ViperStrategy.Slim({
-      // TODO: Do we always remove unfoldings regardless of permission value?
       case unfolding@Unfolding(PredicateAccessPredicate(PredicateAccess(_, name), _), body) =>
-        if (preds(name)) body else unfolding
+        if (cond(name)) body else unfolding
     }, Traverse.BottomUp).execute[Exp](expr)
   }
 
   /**
     * Removes given predicate unfolds and folds from statement.
     *
     * @param stmts A Seqn whose statements will be traversed
-    * @param predIds The set of predicates for which we will remove unfold and fold statements
+    * @param cond The condition a predicate must satisfy to no longer require (un)folding
     * @return The Seqn with all above unfolds and folds removed
     */
-  private[this] def removeFoldUnfolds(stmts: Seqn, predIds: Set[String]): Seqn = {
```

Comment:
This is the fix @Alison-Li and I came up with, not sure if your changes still carry it through.

tl;dr we were just taking _all_ the pre/post condition ids, checking if any fold/unfold statements contained them, and removing them if true.

Reply:
Looks like the `pred` function should take care of it, since it's defined as
```Scala
val (prePredIds, postPredIds) = getPrePostPredIds(method, input, inlinePredIds)
val cond = { pred: String => nonrecursivePredIds(pred) }
```

Reply:
I've made it more general: *all* fold/unfold statements will be removed as long as they're not (mutually) recursive. This takes care of cases where the pre-/postconditions don't actually specify the predicates, but they are being used, for instance when you fold up a predicate that isn't ensured in the postconditions. I'll add a test case for that lower down